packages/@sanity/base/src/components/rovingFocus/useRovingFocus.ts

```diff
@@ -0,0 +1,153 @@
+   * Determine what keys to listen to depending on direction
+   */
+  const nextKey = direction === 'horizontal' ? 'ArrowRight' : 'ArrowDown'
+  const prevKey = direction === 'horizontal' ? 'ArrowLeft' : 'ArrowUp'
+
+  /**
+   * Set focusable elements in state
+   */
+  const handleSetElements = useCallback(() => {
+    if (rootElement) {
+      const els = getFocusableElements(rootElement)
+
+      setFocusableElements(els)
+    }
+  }, [rootElement, ...deps])
+
+  /**
+   * Set focused index
+   */
+  const handleFocus = useCallback((index: number) => {
+    setFocusedIndex(index)
+  }, [])
+
+  /**
+   * Handle increment/decrement of focusedIndex
+   */
+  const handleKeyDown = useCallback(
+    (event) => {
+      if (pause) {
+        return
+      }
+
+      if (event.key === prevKey) {
+        event.preventDefault()
+        setFocusedIndex((prevIndex) => {
+          const next = (prevIndex + lastFocusableIndex) % focusableLen
+
+          if (!loop && next === lastFocusableIndex) {
+            return prevIndex
+          }
+
+          return next
+        })
+      }
+
+      if (event.key === nextKey) {
+        event.preventDefault()
+        setFocusedIndex((prevIndex) => {
+          const next = (prevIndex + 1) % focusableLen
+
+          if (!loop && next === 0) {
+            return prevIndex
+          }
+
+          return next
+        })
+      }
+    },
+    [focusableLen, loop, nextKey, pause, prevKey, lastFocusableIndex, ...deps]
+  )
+
+  /**
+   * Set focusable elements on mount
+   */
+  useEffect(() => {
+    handleSetElements()
+  }, [handleSetElements, initialFocus, direction, ...deps])
+
+  /**
+   * Listen to DOM mutations to update focusableElements with latest state
+   */
+  useEffect(() => {
+    const mo = new MutationObserver(handleSetElements)
+
+    if (rootElement) {
+      mo.observe(rootElement, {
+        childList: true,
+        subtree: true,
+        attributeFilter: MUTATION_ATTRIBUTE_FILTER,
+      })
+    }
+
+    return () => {
+      mo.disconnect()
+    }
+  }, [focusableElements, handleSetElements, rootElement, ...deps])
+
+  /**
+   * Set focus on elements in focusableElements depending on focusedIndex
+   */
+  useEffect(() => {
+    ;(focusableElements as HTMLElement[]).forEach((el, index) => {
+      if (index === focusedIndex) {
+        el.setAttribute('tabIndex', '0')
+        el.setAttribute('aria-selected', 'true')
+        el.focus()
+        el.onfocus = () => handleFocus(index)
+        el.onblur = () => handleFocus(-1)
+      } else {
+        el.setAttribute('tabIndex', '-1')
+        el.setAttribute('aria-selected', 'false')
+        el.onfocus = () => handleFocus(index)
+      }
+    })
+
+    if (focusedIndex === -1 && focusableElements) {
+      const initialIndex = initialFocus === 'last' ? lastFocusableIndex : 0
+      focusableElements[initialIndex]?.setAttribute('tabIndex', '0')
+      focusableElements[initialIndex]?.setAttribute('aria-selected', 'true')
+    }
```

Comment:
how would you feel about extracting this part into its own `useEffect` that sets `focusedIndex`? thinking something like this:

```js
useEffect(() => {
  if (focusedIndex === -1 || focusableElements) {
    setFocusedIndex(initialFocus === 'last' ? lastFocusableIndex : 0)
  }
}, [focusableElements, initialFocus, lastFocusableIndex])
```

the benefit here is that you get the re-use the logic from your other effect (though it requires an additional render). might be missing something though lmk!

Reply:
Good idea! However, we can't `setFocusedIndex` in the `useEffect` you're suggesting, because that will make the first/last button take the focus on mount. We only wan't to control the `tabIndex` to make it focusable when you tab to it.